src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs

```diff
@@ -46,81 +45,59 @@ public async ValueTask SynchronizeSolutionAssetsAsync(Checksum solutionChecksum,
                     checksums.Remove(solutionChecksumObject.Checksum);
                     await _assetProvider.SynchronizeAssetsAsync(checksums, cancellationToken).ConfigureAwait(false);
                 }
+            }
 
-                // third and last get direct children for all projects and documents in the solution 
-                await SynchronizeProjectAssets_NoLockAsync(solutionChecksumObject.Projects, cancellationToken).ConfigureAwait(false);
+            // third and last get direct children for all projects and documents in the solution 
+            foreach (var project in solutionChecksumObject.Projects)
+            {
+                var projectStateChecksums = _assetProvider.GetRequiredAsset<ProjectStateChecksums>(project);
+                await SynchronizeProjectAssetsAsync(projectStateChecksums, cancellationToken).ConfigureAwait(false);
```

Comment:
all this became much more strongly typed.  the solution checksum points at the proejct checksums.  from the project checksums we retrieve the project-state-checkum objects.  which we can then use to synchronize that project.  

i'm trying to make things not just take generic checksums, making it clearer what data needs to be passed to what.

Reply:
Furthermore, instead of blocking the entire operation while doing a sync.  We can now interleave syncing a project's worth of info at a time.